src/credentials/examples/GroupDataProviderExample.cpp

```diff
@@ -0,0 +1,654 @@
+                {
+                    // Reuse existing entry
+                    entry = &fabric->keys[i];
+                    break;
+                }
+            }
+            else if (!entry)
+            {
+                // Unused entry
+                entry = &fabric->keys[i];
+            }
+        }
+        if (entry)
+        {
+            entry->key_set_index = keys.key_set_index;
+            entry->policy        = keys.policy;
+            entry->num_keys_used = keys.num_keys_used;
+            memcpy(entry->epoch_keys, keys.epoch_keys, sizeof(keys.epoch_keys[0]) * keys.num_keys_used);
+            entry->in_use = true;
+            return CHIP_NO_ERROR;
+        }
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    CHIP_ERROR GetKeySet(chip::FabricIndex fabric_index, KeySet & keys) override
+    {
+        VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+        Fabric * fabric = GetExistingFabric(fabric_index);
+        // Search for existing keys
+        for (uint16_t i = 0; fabric && i < kKeyEntriesMax; ++i)
+        {
+            KeysEntry & entry = fabric->keys[i];
+            if (entry.in_use && entry.key_set_index == keys.key_set_index)
+            {
+                // Found
+                keys.policy        = entry.policy;
+                keys.num_keys_used = entry.num_keys_used;
+                memcpy(keys.epoch_keys, entry.epoch_keys, sizeof(keys.epoch_keys[0]) * keys.num_keys_used);
+                return CHIP_NO_ERROR;
+            }
+        }
+        return CHIP_ERROR_KEY_NOT_FOUND;
+    }
+
+    CHIP_ERROR RemoveKeySet(chip::FabricIndex fabric_index, uint16_t key_set_index) override
+    {
+        VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+        Fabric * fabric = GetExistingFabric(fabric_index);
+        // Search for existing keys
+        for (uint16_t i = 0; fabric && i < kKeyEntriesMax; ++i)
+        {
+            KeysEntry & entry = fabric->keys[i];
+            if (entry.in_use && entry.key_set_index == key_set_index)
+            {
+                // Found
+                entry.in_use = false;
+                return CHIP_NO_ERROR;
+            }
+        }
+        return CHIP_ERROR_KEY_NOT_FOUND;
+    }
+
+    KeySetIterator * IterateKeySets(chip::FabricIndex fabric_index) override
+    {
+        VerifyOrReturnError(mInitialized, nullptr);
+        return mKeyIterators.CreateObject(*this, GetExistingFabric(fabric_index));
+    }
+
+    void Release(KeysIterator * iterator) { return mKeyIterators.ReleaseObject(iterator); }
+
+private:
+    bool mInitialized = false;
+    Fabric mFabrics[kNumFabrics];
+    BitMapObjectPool<KeysIterator, kIteratorsMax> mKeyIterators;
+    BitMapObjectPool<EndpointIterator, kIteratorsMax> mEndpointIterators;
+    BitMapObjectPool<FabricGroupStateIterator, kIteratorsMax> mFabricStateIterators;
+    BitMapObjectPool<AllGroupStateIterator, kIteratorsMax> mAllStateIterators;
```

Comment:
What is the use case for dynamically allocating iterators?

Reply:
They are not dynamically allocated if they come from a pool. At first an earlier version of this PR had a "single live slot" constraint but since we can't anticipate all possible usage patterns, there is flexibility in the numbers here for now

Reply:
I meant dynamic as in not-on-the-stack.

I think the answer is that the iterator interfaces by definition require allocation.

Makes sense.